gobblin-core/src/main/java/gobblin/writer/commands/TeradataBufferedInserter.java

```diff
@@ -0,0 +1,201 @@
+    if (this.columnNames == null) {
+      initializeForBatch(this.conn, databaseName, table, jdbcEntryData);
+    }
+    this.pendingInserts.add(jdbcEntryData);
+
+    if (this.pendingInserts.size() == this.batchSize) {
+      insertBatch(this.insertPstmtForFixedBatch); //reuse pre-computed Preparedstatement.
+      return;
+    }
+  }
+
+  private void insertBatch(final PreparedStatement pstmt) {
+    Callable<Boolean> insertCall = new Callable<Boolean>() { //Need a Callable interface to be wrapped by Retryer.
+      @Override
+      public Boolean call() throws Exception {
+        for (JdbcEntryData pendingEntry : TeradataBufferedInserter.this.pendingInserts) {
+          int i = 1;
+          for (JdbcEntryDatum datum : pendingEntry) {
+            Object value = datum.getVal();
+            if (value != null) {
+              pstmt.setObject(i, value);
+            }
+            else {
+              // column type is needed for null value insertion
+              pstmt.setNull(i, columnPosSqlTypes.get(i));
+            }
+            i++;
+          }
+          pstmt.addBatch();
+          pstmt.clearParameters();
+        }
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Executing SQL " + pstmt);
+        }
+        int[] execStatus = pstmt.executeBatch();
+        // Check status explicitly if driver continues batch insertion upon failure 
+        for (int status : execStatus) {
+          if (status == Statement.EXECUTE_FAILED) {
+            throw new BatchUpdateException("Batch insert failed.", execStatus);
+          }
+        }
+        return true;
+      }
+    };
+
+    try {
+      this.retryer.wrap(insertCall).call();
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to insert.", e);
+    }
+    resetBatch();
+  }
+
+  /**
+   * Initializes variables for batch insert and pre-compute PreparedStatement based on requested batch size and parameter size.
+   * @param conn
+   * @param databaseName
+   * @param table
+   * @param jdbcEntryData
+   * @throws SQLException
+   */
+  private void initializeForBatch(Connection conn, String databaseName, String table, JdbcEntryData jdbcEntryData)
+      throws SQLException {
+    this.columnNames = Lists.newArrayList();
+    for (JdbcEntryDatum datum : jdbcEntryData) {
+      this.columnNames.add(datum.getColumnName());
+    }
+    this.pendingInserts = Lists.newArrayList();
+
+    this.insertStmtPrefix =
+        String.format(INSERT_STATEMENT_PREFIX_FORMAT, databaseName, table, JOINER_ON_COMMA.join(this.columnNames));
+    this.insertPstmtForFixedBatch = conn.prepareStatement(createPrepareStatementStr(this.insertStmtPrefix));
+    this.columnPosSqlTypes = getColumnPosSqlTypes();
+    LOG.info(String.format("Initialized for %s insert " + this, (this.batchSize > 1) ? "batch" : ""));
+
+    //retry after 2, 4, 8, 16... sec, max 30 sec delay
+    this.retryer = RetryerBuilder.<Boolean> newBuilder().retryIfException()
```

Comment:
Having a parameter for maxattempt and maxwait? I'll then adapt `MySqlBufferedInserter` as well.